arwen/contexts/async_local.go

```diff
@@ -0,0 +1,454 @@
+}
+
+func (context *asyncContext) createGroupCallbackInput(group *arwen.AsyncCallGroup) *vmcommon.ContractCallInput {
+	runtime := context.host.Runtime()
+
+	input := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallType:       vm.AsynchronousCallBack,
+			CallerAddr:     context.callerAddr,
+			Arguments:      [][]byte{group.CallbackData},
+			CallValue:      big.NewInt(0),
+			GasPrice:       context.gasPrice,
+			GasProvided:    group.GasLocked + context.gasAccumulated,
+			CurrentTxHash:  runtime.GetCurrentTxHash(),
+			OriginalTxHash: runtime.GetOriginalTxHash(),
+			PrevTxHash:     runtime.GetPrevTxHash(),
+		},
+		RecipientAddr: runtime.GetSCAddress(),
+		Function:      group.Callback,
+	}
+
+	logAsync.Trace("created group callback input", "group", group.Identifier, "function", input.Function)
+	logAsync.Trace("created group callback input gas", "provided", input.GasProvided, "locked", group.GasLocked, "accumulated", context.gasAccumulated)
+	return input
+}
+
+func (context *asyncContext) createContextCallbackInput() *vmcommon.ContractCallInput {
+	host := context.host
+	runtime := host.Runtime()
+
+	_, arguments, err := host.CallArgsParser().ParseData(string(context.returnData))
+	if err != nil {
+		arguments = [][]byte{context.returnData}
+	}
+
+	// TODO ensure a new value for VMInput.CurrentTxHash
+	input := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:     context.callerAddr,
+			Arguments:      arguments,
+			CallValue:      runtime.GetVMInput().CallValue,
+			CallType:       vm.AsynchronousCallBack,
+			GasPrice:       runtime.GetVMInput().GasPrice,
+			GasProvided:    context.gasAccumulated,
+			CurrentTxHash:  runtime.GetCurrentTxHash(),
+			OriginalTxHash: runtime.GetOriginalTxHash(),
+			PrevTxHash:     runtime.GetPrevTxHash(),
+		},
+		RecipientAddr: runtime.GetSCAddress(),
+		Function:      context.callback,
+	}
+
+	logAsync.Trace("created context callback input", "sc", runtime.GetSCAddress(), "function", input.Function)
+	logAsync.Trace("created context callback input gas", "provided", input.GasProvided, "accumulated", context.gasAccumulated)
+	return input
+}
+
+func (context *asyncContext) isESDTTransferOnReturnDataWithNoAdditionalData(
+	sndAddr, dstAddr []byte,
+	destinationVMOutput *vmcommon.VMOutput,
+) (bool, string, [][]byte) {
+	if len(destinationVMOutput.ReturnData) == 0 {
+		return false, "", nil
+	}
+
+	argParser := context.host.CallArgsParser()
+	functionName, args, err := argParser.ParseData(string(destinationVMOutput.ReturnData[0]))
+	if err != nil {
+		return false, "", nil
+	}
+
+	return context.isESDTTransferOnReturnDataFromFunctionAndArgs(sndAddr, dstAddr, functionName, args)
+}
+
+func (context *asyncContext) isESDTTransferOnReturnDataFromFunctionAndArgs(
+	sndAddr, dstAddr []byte,
+	functionName string,
+	args [][]byte,
+) (bool, string, [][]byte) {
+	parsedTransfer, err := context.host.ParseESDTTransfers(sndAddr, dstAddr, functionName, args)
```

Comment:
push esdtTransferParser to the asyncContext. don't use it from host

Reply:
Done.